### PR TITLE
I770 Update ImportSource when removing/replacing/taking imported items

### PR DESCRIPTION
**Pull request by kerimoyle**
Addresses #770, #771, #772.
This PR is aimed at fixing how the import sources are updated when imported items are passed around the place. There are a few bugs coming from this single source, hence the three issues above.

### CODE REVIEW

**Comment by kerimoyle** [comment]
There's a bug in the `Parser` (see #771 ) which is highlighted only when the other bugs around the `ImportSource` are fixed.  That one will probably have to be done as part of this PR too, since it's only manifest when these other things are resolved.

**Comment by kerimoyle** [comment]
Ready for review please @agarny @hsorby @nickerso (build on MacOS only failing because of waiting for #760 to be merged).

File: src/api/libcellml/model.h [hunk 1]
```diff
@@ -350,16 +350,20 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
      * Add import source by reference to the model. If the import
      * source item was previously in a different model, it is moved
      * to this one, and the previous model's import source list is updated.
-     *
+     * if the optional @p merge parameter is true, the contents of the given @p importSrc
```
Comment (hsorby) [inline src/api/libcellml/model.h]: If

File: src/componententity.cpp [hunk 2]
```diff
@@ -222,17 +252,25 @@ ComponentPtr ComponentEntity::takeComponent(const std::string &name, bool search
...
+        auto model = owningModel(oldComponent);
+        if (model != nullptr) {
+            model->addImportSource(newComponent->importSource());
+        }
+    }
+
     if (removeComponent(index)) {
-        mPimpl->mComponents.insert(mPimpl->mComponents.begin() + int64_t(index), component);
-        component->setParent(parent);
+        mPimpl->mComponents.insert(mPimpl->mComponents.begin() + int64_t(index), newComponent);
```
Comment (hsorby) [inline src/componententity.cpp:271]: Should this `int64_t` be `uintptr_t`?

**Review by hsorby — DISMISSED** [review]
All tests that go `replaceImportedComponentByNameIndexOrReference` should be separate tests `replaceImportedComponentByName`; `replaceImportedComponentByIndex`; `replaceImportedComponentByReference`.

File: src/componententity.cpp (hunk 2 above)
Comment (kerimoyle) [reply src/componententity.cpp:271]: Perhaps, but if so should be a separate PR as it's everywhere ... and not really related to this PR.  I just changed the name of the component handle so it was easier to understand. 
image

File: src/api/libcellml/model.h (hunk 1 above)
Comment (kerimoyle) [reply src/api/libcellml/model.h]: b82e97c1

**Comment by kerimoyle** [comment]
Review comments all addressed in b82e97c1.  Ready for review please @hsorby @agarny @nickerso.

**Comment by hsorby** [comment]
During my review of this PR I thought that we should reimagine how ImportSources are considered.  I have started a [mini-project](https://github.com/cellml/libcellml/projects/5) to capture this possible transition.

**Pull request merged**